agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -0,0 +1,122 @@
+require_relative '../service_pods/creator'
+require_relative '../service_pods/starter'
+require_relative '../service_pods/stopper'
+require_relative '../service_pods/terminator'
+
+module Kontena::Workers
+  class ServicePodWorker
+    include Celluloid
+    include Kontena::Logging
+    include Kontena::ServicePods::Common
+    include Kontena::Helpers::RpcHelper
+
+    attr_reader :node, :prev_state, :service_pod
+    attr_accessor :service_pod
+
+    def initialize(node, service_pod)
+      @node = node
+      @service_pod = service_pod
+      @prev_state = 'unknown'
+    end
+
+    # @param [Kontena::Models::ServicePod] service_pod
+    def update(service_pod)
+      @service_pod = service_pod
+      ensure_desired_state
+    end
+
+    def destroy
+      @service_pod.mark_as_terminated
+      ensure_desired_state
+    end
+
+    def ensure_desired_state
+      exclusive {
+        debug "state of #{service_pod.name}: #{service_pod.desired_state}"
+        service_container = get_container(service_pod.service_id, service_pod.instance_number)
+        if service_pod.running? && service_container.nil?
+          info "creating #{service_pod.name}"
+          ensure_running
+        elsif service_container && service_pod.running? && !service_container.running?
+          info "starting #{service_pod.name}"
+          ensure_started
+        elsif service_pod.running? && (service_container && service_container_outdated?(service_container, service_pod))
+          info "re-creating #{service_pod.name}"
+          ensure_running
+        elsif service_pod.stopped? && (service_container && service_container.running?)
+          info "stopping #{service_pod.name}"
+          ensure_stopped
+        elsif service_pod.terminated?
+          info "terminating #{service_pod.name}"
+          ensure_terminated if service_container
+        elsif service_pod.desired_state_unknown?
+          info "desired state is unknown for #{service_pod.name}, not doing anything"
+        else
+          warn "unknown state #{service_pod.desired_state} for #{service_pod.name}"
```

Comment:
Oops, looks like this now also needs a "service_pod and container are in sync" (`service_pod.running? && service_container.running?` etc.) case...

```
kontena-agent | W, [2017-03-17T14:56:37.575027 #1]  WARN -- Kontena::Workers::ServicePodWorker: unknown state running for redis-1
```